### ISSUE TITLE
Vec<T>/Vec benchmark

### ISSUE THREAD

**Issue by buybackoff** [Spreads/Spreads.Native#2]
# [`Vec<T>/Vec`](https://github.com/Spreads/Spreads.Native/blob/master/dotnet/src/Spreads.Native/Vec.cs) benchmark

Benchmarking is hard... My casual benchmark initially showed very different results, but `T[]` bound check was definitely eliminated and probably Span's as well. Then in https://github.com/Spreads/Spreads.Native/commit/b861533859080d8c724620a4181fcd92374c9a46 `Vec<T>` was faster.

## **Spreads.Benchmark.Run() results** 

I usually do not use BenchmarkDotNet but run my casual [benchmark helper](https://github.com/Spreads/Spreads.Native/blob/master/dotnet/tests/Spreads.Native.Tests/Benchmark.cs) that measures time inside `using(){}` block like this:

```
                using (Benchmark.Run("VecT", count * mult))
                {
                    for (int m = 0; m < mult; m++)
                    {
                        var z = count - 1;
                        for (int j = 1; j < z; j++)
                        {
                            sum += vecT.GetUnchecked(j - 1);
                        }
                    }
                }
```

This bench measures running time of code regions inside a mega-method (e.g. unit test with many loops and benchmarks). All code is in the same scope for JIT and is run circularly for several cases so probably order impact is eliminated and JIT optimizations are same. Often code from one benchmark affects code from another, as was the case if I uncomment `Vec.Get<T>` case. However, with only three of `T[]/Span<T>/Vec<T>` the numbers are stable and stay the same if running each case alone.

In most cases such casual run already gives a big picture and a single-digit performance difference is not important by itself for my needs. BenchmarkDotNet requires quite a lot of ceremony compared to this simple helper and even with `ShortRun` runs longer than dotTrace or dotMemory. This time is better spent analyzing the profilers output. Such simple benchmark is especially useful with dotTrace unit test integration - I could just comment out what I'm not interested in and quickly profile a unit test. Then when I see that unit test runner distorts the result or I need memory profiling beyond GC counts I just run a particular unit test in a standalone console app (as in this example) and use `R#->Profile->Run startup project performance or memory profiling`.

The results are repeatable by running [`Spreads.Native.Tests.VecTests.ForEachBench`](https://github.com/Spreads/Spreads.Native/blob/46b0dd20f800751aab1a66f5019ea743ce76cafb/dotnet/tests/Spreads.Native.Tests/VecTests.cs#L38-L135) from [console app](https://github.com/Spreads/Spreads.Native/blob/46b0dd20f800751aab1a66f5019ea743ce76cafb/dotnet/tests/Spreads.Native.Run/Program.cs#L32-L33) without debugger)

**ForEachBench**

 Case                |    MOPS |  Elapsed |   GC0 |   GC1 |   GC2 |  Memory
------               |--------:|---------:|------:|------:|------:|--------:
VecT                 | 1647.45 |   607 ms |   0.0 |   0.0 |   0.0 | 0.000 MB
Span                 | 1567.40 |   638 ms |   0.0 |   0.0 |   0.0 | 0.000 MB
Array                | 1567.40 |   638 ms |   0.0 |   0.0 |   0.0 | 0.000 MB

## BenchmarkDotNet results

After I [claimed on Twitter](https://twitter.com/buybackoff/status/1085975263203209216) that `Vec<T>` is faster I felt uncomfortable and I tried to run gold-standard Benchmark.Net, but numbers were clearly nonsensical at the beginning. It looks like bound checks elimination works quite well and at 2-3AM it was not straightforward to break it quickly without killing overall performance. It looks like adding `if` that depends on a `for` var and then accessing array indexer with an index depending on the `for` var is enough, even if this is as simple as:

```
                for (int j = 1; j < _count; j++)
                {
                    if (j >= 42)
                    {
                        sum += arr[j - 1];
                    }
                    else
                    {
                        sum += arr[j];
                    }
                }
```
Also `_count` is `volatile`. I thought that volatile should be enough and maybe it is, but with `if` block results look more like apples to apples. 

The [Benchmark](https://github.com/Spreads/Spreads.Native/blob/46b0dd20f800751aab1a66f5019ea743ce76cafb/dotnet/tests/Spreads.Native.Run/Benchmark.cs#L14) output:

![image](https://user-images.githubusercontent.com/489852/51375501-190d2380-1b17-11e9-9ede-ef09002f912a.png)

_Note how dismal is performance of `Memory<T>.Span[i]` if we were using is as a field of class but require random access._

## Bottom line

Logically `Vec<T>` is doing exactly what `Span<T>` does and the measured method explicitly excludes bound checks - they all must be very close by construction. Fast `Span` is slightly faster in BDN on platforms that support Span natively, slightly slower in my casual benchmark for some reason. `Vec<T>` is on par with `T[]` in BDN (excluding `.NET Core 2.1` where `T[]` is *much* slower, but it looks suspicious) and slightly faster in my bench.

**Overall they are all the same, the difference is not important because absolute performance is so fast that even a single virtual call added to the code path will completely dominate (i.e. kill) the performance.** 

There are few things that make `Vec<T>` better for our use case:
* Could work with native memory - same as `Span<T>`.
* Fast slicing without allocations - same as `Span<T>`.
* Could be a field of a class - same as `T[]`. This is 100% use case so far in Spreads, underlying `T[]` will always be GC-rooted or pinned native memory is used.
* `Vec<T>.Span` property exposes it as `Span<T>` so any API accepting span is available to use.
* Non-generic `Vec` counterpart of `Vec<T>` has exactly the same binary layout and is ideal for columns storage of DataFrame-like structures (which could work on native memory, e.g. we could use R-owned memory). `Vec.Get<T>` is as fast as `Vec<T>.Get` and converting between the two is fast and trivial.
* Non-generic `Vec` setter accepts objects but stores info about underlying type `T` and uses `dynamic`, so passing `int` to a `Vec` of `long`s just works. Performance impact of `dynamic` is quite tolerable and flexibility benefits are much greater. `Vec` getter/setter redirect to `Vec<T>` getter/setter via ldftn + calli and this is the dominating cost vs `Vec<T>` all-inlined path.

I really hope that _coreclr_ adds `non-ref` counterpart of `Span<T>` rather sooner (3.0 maybe?) because it absence (before reinventing `Vec<T>/Vec`) blocked important use case for fast data analysis.

cc @adamsitnik @jkotas

P.S. I have an important question left for _coreclr_ gurus: ELI5 what was wrong with slice.net `Slice<T>` being non `ref struct` other than potential GC problems? Is it safe to use `Vec<T>/Slice<T>` if the underlying array is always GC-rooted? IL code [is here](https://github.com/Spreads/Spreads.Native/blob/46b0dd20f800751aab1a66f5019ea743ce76cafb/dotnet/src/Spreads.Native/UnsafeEx.il#L362-L387).

**Comment by buybackoff** [comment]
@kkokosa you wanted to take a look and you are probably the best to answer the question in P.S. above since it is about memory/GC :)

**Comment by jkotas** [comment]
> I really hope that coreclr adds non-ref counterpart of Span<T> rather sooner (3.0 maybe)

There is no [CoreFX API proposal](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/api-review-process.md) like this as far as I know. It means that this is not happening anytime soon.

> Is it safe to use `Vec<T>/Slice<T>` if the underlying array is always GC-rooted? IL code is here

The IL code you have linked looks like invalid IL to me. You can run your tests with debug build of RyuJIT. I expect you will see asserts in RyuJIT and the generated code to be not always correct.

**Comment by buybackoff** [comment]
@jkotas the IL code is from slice ndt, I assumed the author knew what he
was doing and there were comments about GC is happy with that. How to fix
the IL so that it works correctly?

**Comment by jkotas** [comment]
What is slice ntd?

**Comment by buybackoff** [comment]
Sorry for typo, slice dot net library by Joe Duffy, old Span prototype.
https://github.com/joeduffy/slice.net

**Comment by jkotas** [comment]
Yes, it was a know problem with that prototype. There was a reason why we just have not taken this code and shipped it as portable Span. I guess you would need to do something along the lines what the shipping slow Span did.

**Comment by buybackoff** [comment]
@jkotas 
> Yes, it was a know problem with that prototype. There was a reason why we just have not taken this code and shipped it as portable Span. I guess you would need to do something along the lines what the shipping slow Span did.

Does this mean that the code in `corefxlab` just before Span was removed from there was also incorrect? Here is the link to the last version in corefxlab: https://github.com/dotnet/corefxlab/blob/c9f9a016f37cee4b9b9fe07729180b110502b863/src/System.Slices/System/UnsafeUtilities.cs#L78-L114

I have added this and Vec<T> has become significantly faster (instead of the previous version where we calculated offset in C# and passed it to the same first-half-of-the-method with 2 args)
```
	.method hidebysig static !!T GetAtIndex<T>(
		object obj, native int offset, native int index) cil managed aggressiveinlining
	{
		.maxstack 3
        .locals([0] uint8 & addr)
        ldarg.0     // load the object
        stloc.0     // convert the object pointer to a byref
        ldloc.0     // load the object pointer as a byref
        ldarg.1     // load the offset
        add         // add the offset
        ldarg.2     // load the index
        sizeof !!T  // load size of T
        mul         // multiply the index and size of T
        add         // add the result
        ldobj !!T   // load a T value from the computed address
        ret
	}

```
Currently looking in the issues and corefx repo history on what happened after that. 

But what is exactly wrong with this IL?

**Comment by jkotas** [comment]
> Does this mean that the code in corefxlab just before Span was removed from there was also incorrect? 

Yes, the original corefxlab code was not valid IL. See https://github.com/dotnet/corefxlab/pull/796#issuecomment-246667113 .
